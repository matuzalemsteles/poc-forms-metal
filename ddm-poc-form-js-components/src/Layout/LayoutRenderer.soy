{namespace LayoutRenderer}

{template .render}
    {@param spritemap: string}
    {@param? _handleDeleteButtonClicked: any}
    {@param? _handleFieldChange: any}
    {@param? _handleFocusSelectField: any}
    {@param? _handleOnClickResize: any}
    {@param? activePage: number}
    {@param? editable: bool}
    {@param? modeRenderer: string}
    {@param? pages: ?}

    {let $contentRenderer: $modeRenderer ?: 'grid' /}

    {delcall LayoutRenderer.pages variant="$contentRenderer"}
        {param _handleDeleteButtonClicked: $_handleDeleteButtonClicked /}
        {param _handleFieldChange: $_handleFieldChange /}
        {param _handleFocusSelectField: $_handleFocusSelectField /}
        {param _handleOnClickResize: $_handleOnClickResize /}
        {param activePage: $activePage /}
        {param contentRenderer: $contentRenderer /}
        {param editable: $editable /}
        {param pages: $pages /}
        {param spritemap: $spritemap /}
    {/delcall}
{/template}

{deltemplate LayoutRenderer.pages variant="'grid'"}
    {@param spritemap: string}
    {@param? _handleDeleteButtonClicked: any}
    {@param? _handleFieldChange: any}
    {@param? _handleFocusSelectField: any}
    {@param? _handleOnClickResize: any}
    {@param? activePage: number}
    {@param? contentRenderer: string}
    {@param? editable: bool}
    {@param? pages: ?}

    {for $page in $pages}
        <div class="form-builder-layout">
            {delcall LayoutRenderer.page variant="$contentRenderer"}
                {param _handleDeleteButtonClicked: $_handleDeleteButtonClicked /}
                {param _handleFieldChange: $_handleFieldChange /}
                {param _handleFocusSelectField: $_handleFocusSelectField /}
                {param _handleOnClickResize: $_handleOnClickResize /}
                {param contentRenderer: $contentRenderer /}
                {param editable: $editable /}
                {param page: $page /}
                {param pageId: index($page) /}
                {param spritemap: $spritemap /}
            {/delcall}
        </div>
    {/for}
{/deltemplate}

{deltemplate LayoutRenderer.pages variant="'list'"}
    {@param spritemap: string}
    {@param? _handleDeleteButtonClicked: any}
    {@param? _handleFieldChange: any}
    {@param? _handleFocusSelectField: any}
    {@param? _handleOnClickResize: any}
    {@param? activePage: number}
    {@param? contentRenderer: string}
    {@param? editable: bool}
    {@param? pages: ?}

    {for $page in $pages}
        {let $classes kind="text"}
            fade tab-pane
            {if $activePage == index($page)}
                {sp}active show
            {/if}
        {/let}

        <div aria-labelledby="sidebarLightFormsFieldTab" class="{$classes}" id="sidebarLightDetails" role="tabpanel">
            {delcall LayoutRenderer.page variant="$contentRenderer"}
                {param _handleDeleteButtonClicked: $_handleDeleteButtonClicked /}
                {param _handleFieldChange: $_handleFieldChange /}
                {param _handleFocusSelectField: $_handleFocusSelectField /}
                {param _handleOnClickResize: $_handleOnClickResize /}
                {param contentRenderer: $contentRenderer /}
                {param editable: $editable /}
                {param page: $page /}
                {param pageId: index($page) /}
                {param spritemap: $spritemap /}
            {/delcall}
        </div>
    {/for}
{/deltemplate}

{deltemplate LayoutRenderer.page variant="'list'"}
    {@param spritemap: string}
    {@param? _handleDeleteButtonClicked: any}
    {@param? _handleFieldChange: any}
    {@param? _handleFocusSelectField: any}
    {@param? _handleOnClickResize: any}
    {@param? contentRenderer: string}
    {@param? editable: bool}
    {@param? page: ?}
    {@param? pageId: number}
    
    {for $row in $page.rows}
        {delcall LayoutRenderer.row variant="'grid'"}
            {param _handleDeleteButtonClicked: $_handleDeleteButtonClicked /}
            {param _handleFieldChange: $_handleFieldChange /}
            {param _handleFocusSelectField: $_handleFocusSelectField /}
            {param _handleOnClickResize: $_handleOnClickResize /}
            {param contentRenderer: $contentRenderer /}
            {param editable: $editable /}
            {param indexRow: index($row) /}
            {param pageId: $pageId /}
            {param row: $row /}
            {param spritemap: $spritemap /}
        {/delcall}
    {/for}
{/deltemplate}

{deltemplate LayoutRenderer.page variant="'grid'"}
    {@param spritemap: string}
    {@param? _handleDeleteButtonClicked: any}
    {@param? _handleFieldChange: any}
    {@param? _handleFocusSelectField: any}
    {@param? _handleOnClickResize: any}
    {@param? contentRenderer: string}
    {@param? editable: bool}
    {@param? page: ?}
    {@param? pageId: number}

    <div class="layout-page">
        {for $row in $page.rows}
            {if isFirst($row) and $editable}
                {call .fieldEmpty}
                    {param isRow: true /}
                    {param pageId: $pageId /}
                    {param row: 0 /}
                    {param size: 12 /}
                {/call}
            {/if}

            {delcall LayoutRenderer.row variant="$contentRenderer"}
                {param _handleDeleteButtonClicked: $_handleDeleteButtonClicked /}
                {param _handleFieldChange: $_handleFieldChange /}
                {param _handleFocusSelectField: $_handleFocusSelectField /}
                {param _handleOnClickResize: $_handleOnClickResize /}
                {param contentRenderer: $contentRenderer /}
                {param editable: $editable /}
                {param indexRow: index($row) /}
                {param pageId: $pageId /}
                {param row: $row /}
                {param spritemap: $spritemap /}
            {/delcall}

            {if $editable}
                {call .fieldEmpty}
                    {param isRow: true /}
                    {param pageId: $pageId /}
                    {param row: index($row) + 1 /}
                    {param size: 12 /}
                {/call}
            {/if}
        {/for}
    </div>
{/deltemplate}

{deltemplate LayoutRenderer.row variant="'grid'"}
    {@param spritemap: string}
    {@param? _handleDeleteButtonClicked: any}
    {@param? _handleFieldChange: any}
    {@param? _handleFocusSelectField: any}
    {@param? _handleOnClickResize: any}
    {@param? contentRenderer: string}
    {@param? editable: bool}
    {@param? indexRow: number}
    {@param? pageId: number}
    {@param? row: ?}

    <div class="row">
        {for $column in $row.columns}
            {delcall LayoutRenderer.column variant="$contentRenderer"}
                {param _handleDeleteButtonClicked: $_handleDeleteButtonClicked /}
                {param _handleFieldChange: $_handleFieldChange /}
                {param _handleFocusSelectField: $_handleFocusSelectField /}
                {param _handleOnClickResize: $_handleOnClickResize /}
                {param column: $column /}
                {param editable: $editable /}
                {param indexColumn: index($column) /}
                {param indexRow: $indexRow /}
                {param pageId: $pageId /}
                {param spritemap: $spritemap /}
            {/delcall}
        {/for}
    </div>
{/deltemplate}

{deltemplate LayoutRenderer.column variant="'list'"}
    {@param spritemap: string}
    {@param? _handleDeleteButtonClicked: any}
    {@param? _handleFieldChange: any}
    {@param? _handleFocusSelectField: any}
    {@param? _handleOnClickResize: any}
    {@param? column: ?}
    {@param? editable: bool}
    {@param? indexColumn: number}
    {@param? indexRow: number}
    {@param? pageId: number}

    {if length($column.fields) != 0}
        {let $attributes kind="attributes"}
            class="col-md-{$column.size}"
        {/let}

        <div {$attributes}>
            {for $field in $column.fields}
                {let $fieldType: $field.type ?: 'empty' /}

                {delcall LayoutRenderer.RegisterFieldType variant="$fieldType" data="$field" allowemptydefault="true"}
                    {param editable: $editable /}
                    {param events: [
                            'fieldEdit': $_handleFieldChange
                    ] /}
                    {param spritemap: $spritemap /}
                {/delcall}
            {/for}
        </div>
    {/if}
{/deltemplate}

{deltemplate LayoutRenderer.column variant="'grid'"}
    {@param spritemap: string}
    {@param? _handleDeleteButtonClicked: any}
    {@param? _handleFieldChange: any}
    {@param? _handleFocusSelectField: any}
    {@param? _handleOnClickResize: any}
    {@param? column: ?}
    {@param? editable: bool}
    {@param? indexColumn: number}
    {@param? indexRow: number}
    {@param? pageId: number}

    {if length($column.fields) != 0}
        {let $attributes kind="attributes"}
            class="col-md-{$column.size} col-ddm"

            {if $indexRow != 'undefined'}
                data-ddm-field-row="{$indexRow}"
            {/if}

            {if $pageId != 'undefined'}
                data-ddm-field-page="{$pageId}"
            {/if}

            {if $indexColumn != 'undefined'}
                data-ddm-field-column="{$indexColumn}"
            {/if}
        {/let}

        <div {$attributes}>
            {let $content kind="html"}
                {for $field in $column.fields}
                    {let $fieldType: $field.type /}

                    {delcall LayoutRenderer.RegisterFieldType variant="$fieldType" data="$field" allowemptydefault="true"}
                        {param editable: $editable /}
                        {param events: [
                            'fieldEdit': $_handleFieldChange
                        ] /}
                        {param ref: 'field' /}
                        {param spritemap: $spritemap /}
                    {/delcall}
                {/for}
            {/let}

            {if $editable}
                <div class="ddm-field-container">
                    <div class="ddm-resize-handle-left" data-onclick="{$_handleOnClickResize}"></div> 

                    {let $attributesDrag kind="attributes"}
                        class="ddm-drag"

                        data-onclick="{$_handleFocusSelectField}"
                    {/let}

                    <div {$attributesDrag}>
                        {$content}
                    </div>

                    {call .fieldActions}
                        {param _handleDeleteButtonClicked: $_handleDeleteButtonClicked /}
                        {param spritemap: $spritemap /}
                    {/call}

                    <div class="ddm-resize-handle-right" data-onclick="{$_handleOnClickResize}"></div>
                </div>
            {else}
                {$content}
            {/if}
        </div>
    {elseif $editable}
        {call .fieldEmpty}
            {param column: $indexColumn /}
            {param pageId: $pageId /}
            {param row: $indexRow /}
            {param size: $column.size /}
        {/call}
    {/if}
{/deltemplate}

{template .fieldActions}
    {@param spritemap: string}
    {@param? _handleDeleteButtonClicked: any}

    <div class="ddm-field-actions-container">
        {call ClayButton.render}
            {param style: 'secondary' /}
            {param size: 'sm' /}
            {param spritemap: $spritemap /}
            {param icon: 'paste' /}
            {param disabled: true /}
            {param monospaced: true /}
        {/call}
        {call ClayButton.render}
            {param events: [
                'click': $_handleDeleteButtonClicked
            ] /}

            {param size: 'sm' /}
            {param style: 'secondary' /}
            {param spritemap: $spritemap /}
            {param icon: 'trash' /}
            {param monospaced: true /}
        {/call}
    </div>
{/template}

{template .fieldEmpty}
    {@param size: number}
    {@param? column: number}
    {@param? isRow: bool}
    {@param? pageId: number}
    {@param? row: number}

    {let $attributes kind="attributes"}
        class="col-md-{$size}" 

        {if $row != 'undefined'}
            data-ddm-field-row="{$row}"
        {/if}

        {if $pageId != 'undefined'}
            data-ddm-field-page="{$pageId}"
        {/if}

        {if $column != 'undefined'}
            data-ddm-field-column="{$column}"
        {/if}
    {/let}

    {let $content kind="html"}
        <div {$attributes}>
            <div class="ddm-target col-empty"></div>
        </div>
    {/let}

    {if $isRow}
        <div class="row">
            {$content}
        </div>
    {else}
        {$content}
    {/if}
{/template}

{deltemplate LayoutRenderer.RegisterFieldType}
    {@param? editable: bool}
    {@param? field: ?}

    {if $editable}
        <div>
            Field doesn't exist.
        </div>
    {/if}
{/deltemplate}