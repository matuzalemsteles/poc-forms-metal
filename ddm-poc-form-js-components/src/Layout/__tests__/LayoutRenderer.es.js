import Context from './__mock__/mockContext.es';
import Fields from './__fixtures__/Fields.es';
import LayoutRenderer from '../LayoutRenderer.es';
import {dom as MetalTestsUtil} from 'metal-dom';
import LayoutSupport from '../LayoutSupport.es';

let component;
let context = null;
let spritemap = 'icons.svg';

describe('LayoutRenderer', () => {
    beforeEach(() => {
        context = JSON.parse(JSON.stringify(Context));

        jest.useFakeTimers();
    });

    afterEach(() => {
        if (component) {
			component.dispose();
        }

        context = null;
    });

    it('should render default markup', () => {
        component = new LayoutRenderer({
            spritemap,
        });

        expect(component).toMatchSnapshot();
    });

    it('should render a layout with pages', () => {
        component = new LayoutRenderer({
            pages: context,
            spritemap,
        });

        expect(component).toMatchSnapshot();
    });

    it('should render a layout with pages in mode of list', () => {
        component = new LayoutRenderer({
            pages: context,
            modeRenderer: 'list',
            spritemap,
        });

        expect(component).toMatchSnapshot();
    });

    it('should render a layout in editable mode', () => {
        component = new LayoutRenderer({
            pages: context,
            editable: true,
            spritemap,
        });

        expect(component).toMatchSnapshot();
    });

    it('should render a layout with disabled drag and drop', () => {
        component = new LayoutRenderer({
            disabledDragAndDrop: true,
            editable: true,
            pages: context,
            spritemap
        });

        expect(component._dragAndDrop).toBeUndefined();
    });

    it('should render a layout and emit an event on button delete clicked', () => {
        component = new LayoutRenderer({
            pages: context,
            editable: true,
            spritemap
        });

        const spy = jest.spyOn(component, 'emit');
        
        component.element.querySelector("button[aria-label='trash']").click();

        expect(spy).toHaveBeenCalled();
        expect(spy).toHaveBeenCalledWith(
			'deleteButtonClicked',
			expect.any(Object)
		);
    });

    it('should render a layout and continue to propagate the field edit event', () => {
        component = new LayoutRenderer({
            pages: context,
            editable: true,
            spritemap
        });

        const spy = jest.spyOn(component, 'emit');

        component.refs.field.emitFieldEdit();

        expect(spy).toHaveBeenCalled();
        expect(spy).toHaveBeenCalledWith(
			'fieldEdit',
			{
                value: 'Foo',
                key: 'Bar',
                originalEvent: {}
            }
		);
    });

    it('should render a layout and emit the field move event', () => {
        component = new LayoutRenderer({
            pages: context,
            editable: true,
            spritemap
        });

        const spy = jest.spyOn(component, 'emit');

        const target = component.element.querySelector('.ddm-target');
        const source = component.element.querySelectorAll('.ddm-drag').item(2);

        const mockEvent = {
            target,
            source
        };

        component._handleDragAndDropEnd(mockEvent);

        expect(spy).toHaveBeenCalled();
        expect(spy).toHaveBeenCalledWith(
			'fieldMove',
			{
                data: mockEvent,
                target: {
                    indexColumn: false,
                    indexPage: 0,
                    indexRow: 0
                },
                source: {
                    indexColumn: 1,
                    indexPage: 0,
                    indexRow: 1
                }
            }
		);
    });

    it('should render a layout and ignore the field move when there is no target', () => {
        component = new LayoutRenderer({
            pages: context,
            editable: true,
            spritemap
        });

        const spy = jest.spyOn(component, 'emit');

        const source = component.element.querySelectorAll('.ddm-drag').item(2);

        const mockEvent = {
            target: null,
            source
        };

        component._handleDragAndDropEnd(mockEvent);

        expect(spy).not.toHaveBeenCalled();
        expect(spy).not.toHaveBeenCalledWith(
			'fieldMove',
			expect.any(Object)
		);
    });

    it('should render a layout with emit an field clicked event', () => {
        component = new LayoutRenderer({
            disabledDragAndDrop: true,
            editable: true,
            pages: context,
            spritemap
        });

        const spy = jest.spyOn(component, 'emit');

        component.element.querySelector(".ddm-drag").click();

        expect(spy).toHaveBeenCalled();
        expect(spy).toHaveBeenCalledWith(
			'fieldClicked',
			expect.any(Object)
		);
    });

    it('should emit a fieldClicked event with the field location', () => {
        component = new LayoutRenderer({
            disabledDragAndDrop: true,
            editable: true,
            pages: context,
            spritemap
        });

        const spy = jest.spyOn(component, 'emit');

        component.element.querySelector(".ddm-drag").click();

        expect(spy).toHaveBeenCalled();
        expect(spy).toHaveBeenCalledWith(
			'fieldClicked',
			expect.objectContaining({
                indexColumn: expect.anything(),
                indexPage: expect.any(Number),
                indexRow: expect.any(Number)
            })
		);
    });

    it('should render a layout and reset drag and drop to every change of API pages when editable is true', () => {
        component = new LayoutRenderer({
            editable: true,
            pages: context,
            spritemap
        });

        const spy = jest.spyOn(component, '_startDrag');
        const spyDragAndDrop = jest.spyOn(component._dragAndDrop, 'disposeInternal');

        const newContext = LayoutSupport.removeFields(context, 0, 1, 0);

        component.setState({
            pages: newContext
        });

        jest.runAllTimers();

        expect(spy).toHaveBeenCalled();
        expect(spyDragAndDrop).toHaveBeenCalled();
    });

    it('should render a layout and if it is not editable should not reset the drag-and-drop feature for all API page changes', () => {
        component = new LayoutRenderer({
            editable: false,
            pages: context,
            spritemap
        });

        const spy = jest.spyOn(component, '_startDrag');

        const newContext = LayoutSupport.removeFields(context, 0, 1, 0);

        component.setState({
            pages: newContext
        });

        jest.runAllTimers();

        expect(spy).not.toHaveBeenCalled();
    });

    it('should render a layout and if it is disabled should not reset the drag-and-drop feature for all API page changes', () => {
        component = new LayoutRenderer({
            disabledDragAndDrop: true,
            editable: true,
            pages: context,
            spritemap
        });

        const spy = jest.spyOn(component, '_startDrag');

        const newContext = LayoutSupport.removeFields(context, 0, 1, 0);

        component.setState({
            pages: newContext
        });

        jest.runAllTimers();

        expect(spy).not.toHaveBeenCalled();
    });

    it('should render a layout with an empty field only in editable mode', () => {
        const indexColumn = 2;
        const indexPage = 0;
        const indexRow = 1;
        const fields = [
            {
                type: 'option_multiple',
                spritemap: 'icons.svg'
            }
        ];
        const newContext = LayoutSupport.addFields(context, indexPage, indexRow, indexColumn, fields);

        component = new LayoutRenderer({
            editable: true,
            pages: newContext,
            spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should not render the layout with the field empty in non-editable mode', () => {
        const indexColumn = 2;
        const indexPage = 0;
        const indexRow = 1;
        const fields = [
            {
                type: 'option_multiple',
                spritemap: 'icons.svg'
            }
        ];
        const newContext = LayoutSupport.addFields(context, indexPage, indexRow, indexColumn, fields);

        component = new LayoutRenderer({
            editable: false,
            pages: newContext,
            spritemap
        });

        expect(component).toMatchSnapshot();
    });
});