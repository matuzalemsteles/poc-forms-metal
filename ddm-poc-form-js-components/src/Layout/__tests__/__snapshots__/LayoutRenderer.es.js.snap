// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutRenderer should not render the layout with the field empty in non-editable mode 1`] = `
<div class="form-builder-layout">
  <div class="layout-page">
    <div class="row">
      <div class="col-md-3 col-ddm" data-ddm-field-row="0" data-ddm-field-page="0" data-ddm-field-column="0">
        <p>Radio Field</p>
      </div>
    </div>
    <div class="row">
      <div class="col-md-4 col-ddm" data-ddm-field-row="1" data-ddm-field-page="0" data-ddm-field-column="0">
        <p>Text Field</p>
      </div>
      <div class="col-md-6 col-ddm" data-ddm-field-row="1" data-ddm-field-page="0" data-ddm-field-column="1">
        <p>Select Field</p>
      </div>
      <div class="col-md-2 col-ddm" data-ddm-field-row="1" data-ddm-field-page="0" data-ddm-field-column="2"></div>
    </div>
    <div class="row">
      <div class="col-md-12 col-ddm" data-ddm-field-row="2" data-ddm-field-page="0" data-ddm-field-column="0">
        <p>Text Field</p>
      </div>
    </div>
  </div>
</div>
`;

exports[`LayoutRenderer should render a layout in editable mode 1`] = `
<div class="form-builder-layout">
  <div class="layout-page">
    <div class="row">
      <div class="col-md-12" data-ddm-field-row="0" data-ddm-field-page="0">
        <div class="ddm-target col-empty"></div>
      </div>
    </div>
    <div class="row">
      <div class="col-md-3 col-ddm" data-ddm-field-row="0" data-ddm-field-page="0" data-ddm-field-column="0">
        <div class="ddm-field-container">
          <div class="ddm-resize-handle-left"></div>
          <div class="ddm-drag">
            <p>Radio Field</p>
          </div>
          <div class="ddm-field-actions-container">
            <button class="btn btn-monospaced btn-sm btn-secondary" aria-label="paste" disabled="disabled" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-paste" focusable="false">
                <title>paste</title>
                <use xlink:href="icons.svg#paste"></use>
              </svg>
            </button>
            <button class="btn btn-monospaced btn-sm btn-secondary" aria-label="trash" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-trash" focusable="false">
                <title>trash</title>
                <use xlink:href="icons.svg#trash"></use>
              </svg>
            </button>
          </div>
          <div class="ddm-resize-handle-right"></div>
        </div>
      </div>
      <div class="col-md-9" data-ddm-field-row="0" data-ddm-field-page="0" data-ddm-field-column="1">
        <div class="ddm-target col-empty"></div>
      </div>
    </div>
    <div class="row">
      <div class="col-md-12" data-ddm-field-row="1" data-ddm-field-page="0">
        <div class="ddm-target col-empty"></div>
      </div>
    </div>
    <div class="row">
      <div class="col-md-4 col-ddm" data-ddm-field-row="1" data-ddm-field-page="0" data-ddm-field-column="0">
        <div class="ddm-field-container">
          <div class="ddm-resize-handle-left"></div>
          <div class="ddm-drag">
            <p>Text Field</p>
          </div>
          <div class="ddm-field-actions-container">
            <button class="btn btn-monospaced btn-sm btn-secondary" aria-label="paste" disabled="disabled" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-paste" focusable="false">
                <title>paste</title>
                <use xlink:href="icons.svg#paste"></use>
              </svg>
            </button>
            <button class="btn btn-monospaced btn-sm btn-secondary" aria-label="trash" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-trash" focusable="false">
                <title>trash</title>
                <use xlink:href="icons.svg#trash"></use>
              </svg>
            </button>
          </div>
          <div class="ddm-resize-handle-right"></div>
        </div>
      </div>
      <div class="col-md-6 col-ddm" data-ddm-field-row="1" data-ddm-field-page="0" data-ddm-field-column="1">
        <div class="ddm-field-container">
          <div class="ddm-resize-handle-left"></div>
          <div class="ddm-drag">
            <p>Select Field</p>
          </div>
          <div class="ddm-field-actions-container">
            <button class="btn btn-monospaced btn-sm btn-secondary" aria-label="paste" disabled="disabled" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-paste" focusable="false">
                <title>paste</title>
                <use xlink:href="icons.svg#paste"></use>
              </svg>
            </button>
            <button class="btn btn-monospaced btn-sm btn-secondary" aria-label="trash" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-trash" focusable="false">
                <title>trash</title>
                <use xlink:href="icons.svg#trash"></use>
              </svg>
            </button>
          </div>
          <div class="ddm-resize-handle-right"></div>
        </div>
      </div>
      <div class="col-md-2" data-ddm-field-row="1" data-ddm-field-page="0" data-ddm-field-column="2">
        <div class="ddm-target col-empty"></div>
      </div>
    </div>
    <div class="row">
      <div class="col-md-12" data-ddm-field-row="2" data-ddm-field-page="0">
        <div class="ddm-target col-empty"></div>
      </div>
    </div>
    <div class="row">
      <div class="col-md-12 col-ddm" data-ddm-field-row="2" data-ddm-field-page="0" data-ddm-field-column="0">
        <div class="ddm-field-container">
          <div class="ddm-resize-handle-left"></div>
          <div class="ddm-drag">
            <p>Text Field</p>
          </div>
          <div class="ddm-field-actions-container">
            <button class="btn btn-monospaced btn-sm btn-secondary" aria-label="paste" disabled="disabled" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-paste" focusable="false">
                <title>paste</title>
                <use xlink:href="icons.svg#paste"></use>
              </svg>
            </button>
            <button class="btn btn-monospaced btn-sm btn-secondary" aria-label="trash" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-trash" focusable="false">
                <title>trash</title>
                <use xlink:href="icons.svg#trash"></use>
              </svg>
            </button>
          </div>
          <div class="ddm-resize-handle-right"></div>
        </div>
      </div>
    </div>
    <div class="row">
      <div class="col-md-12" data-ddm-field-row="3" data-ddm-field-page="0">
        <div class="ddm-target col-empty"></div>
      </div>
    </div>
  </div>
</div>
`;

exports[`LayoutRenderer should render a layout with an empty field only in editable mode 1`] = `
<div class="form-builder-layout">
  <div class="layout-page">
    <div class="row">
      <div class="col-md-12" data-ddm-field-row="0" data-ddm-field-page="0">
        <div class="ddm-target col-empty"></div>
      </div>
    </div>
    <div class="row">
      <div class="col-md-3 col-ddm" data-ddm-field-row="0" data-ddm-field-page="0" data-ddm-field-column="0">
        <div class="ddm-field-container">
          <div class="ddm-resize-handle-left"></div>
          <div class="ddm-drag">
            <p>Radio Field</p>
          </div>
          <div class="ddm-field-actions-container">
            <button class="btn btn-monospaced btn-sm btn-secondary" aria-label="paste" disabled="disabled" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-paste" focusable="false">
                <title>paste</title>
                <use xlink:href="icons.svg#paste"></use>
              </svg>
            </button>
            <button class="btn btn-monospaced btn-sm btn-secondary" aria-label="trash" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-trash" focusable="false">
                <title>trash</title>
                <use xlink:href="icons.svg#trash"></use>
              </svg>
            </button>
          </div>
          <div class="ddm-resize-handle-right"></div>
        </div>
      </div>
      <div class="col-md-9" data-ddm-field-row="0" data-ddm-field-page="0" data-ddm-field-column="1">
        <div class="ddm-target col-empty"></div>
      </div>
    </div>
    <div class="row">
      <div class="col-md-12" data-ddm-field-row="1" data-ddm-field-page="0">
        <div class="ddm-target col-empty"></div>
      </div>
    </div>
    <div class="row">
      <div class="col-md-4 col-ddm" data-ddm-field-row="1" data-ddm-field-page="0" data-ddm-field-column="0">
        <div class="ddm-field-container">
          <div class="ddm-resize-handle-left"></div>
          <div class="ddm-drag">
            <p>Text Field</p>
          </div>
          <div class="ddm-field-actions-container">
            <button class="btn btn-monospaced btn-sm btn-secondary" aria-label="paste" disabled="disabled" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-paste" focusable="false">
                <title>paste</title>
                <use xlink:href="icons.svg#paste"></use>
              </svg>
            </button>
            <button class="btn btn-monospaced btn-sm btn-secondary" aria-label="trash" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-trash" focusable="false">
                <title>trash</title>
                <use xlink:href="icons.svg#trash"></use>
              </svg>
            </button>
          </div>
          <div class="ddm-resize-handle-right"></div>
        </div>
      </div>
      <div class="col-md-6 col-ddm" data-ddm-field-row="1" data-ddm-field-page="0" data-ddm-field-column="1">
        <div class="ddm-field-container">
          <div class="ddm-resize-handle-left"></div>
          <div class="ddm-drag">
            <p>Select Field</p>
          </div>
          <div class="ddm-field-actions-container">
            <button class="btn btn-monospaced btn-sm btn-secondary" aria-label="paste" disabled="disabled" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-paste" focusable="false">
                <title>paste</title>
                <use xlink:href="icons.svg#paste"></use>
              </svg>
            </button>
            <button class="btn btn-monospaced btn-sm btn-secondary" aria-label="trash" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-trash" focusable="false">
                <title>trash</title>
                <use xlink:href="icons.svg#trash"></use>
              </svg>
            </button>
          </div>
          <div class="ddm-resize-handle-right"></div>
        </div>
      </div>
      <div class="col-md-2 col-ddm" data-ddm-field-row="1" data-ddm-field-page="0" data-ddm-field-column="2">
        <div class="ddm-field-container">
          <div class="ddm-resize-handle-left"></div>
          <div class="ddm-drag">
            <div>Field doesn't exist.</div>
          </div>
          <div class="ddm-field-actions-container">
            <button class="btn btn-monospaced btn-sm btn-secondary" aria-label="paste" disabled="disabled" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-paste" focusable="false">
                <title>paste</title>
                <use xlink:href="icons.svg#paste"></use>
              </svg>
            </button>
            <button class="btn btn-monospaced btn-sm btn-secondary" aria-label="trash" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-trash" focusable="false">
                <title>trash</title>
                <use xlink:href="icons.svg#trash"></use>
              </svg>
            </button>
          </div>
          <div class="ddm-resize-handle-right"></div>
        </div>
      </div>
    </div>
    <div class="row">
      <div class="col-md-12" data-ddm-field-row="2" data-ddm-field-page="0">
        <div class="ddm-target col-empty"></div>
      </div>
    </div>
    <div class="row">
      <div class="col-md-12 col-ddm" data-ddm-field-row="2" data-ddm-field-page="0" data-ddm-field-column="0">
        <div class="ddm-field-container">
          <div class="ddm-resize-handle-left"></div>
          <div class="ddm-drag">
            <p>Text Field</p>
          </div>
          <div class="ddm-field-actions-container">
            <button class="btn btn-monospaced btn-sm btn-secondary" aria-label="paste" disabled="disabled" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-paste" focusable="false">
                <title>paste</title>
                <use xlink:href="icons.svg#paste"></use>
              </svg>
            </button>
            <button class="btn btn-monospaced btn-sm btn-secondary" aria-label="trash" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-trash" focusable="false">
                <title>trash</title>
                <use xlink:href="icons.svg#trash"></use>
              </svg>
            </button>
          </div>
          <div class="ddm-resize-handle-right"></div>
        </div>
      </div>
    </div>
    <div class="row">
      <div class="col-md-12" data-ddm-field-row="3" data-ddm-field-page="0">
        <div class="ddm-target col-empty"></div>
      </div>
    </div>
  </div>
</div>
`;

exports[`LayoutRenderer should render a layout with pages 1`] = `
<div class="form-builder-layout">
  <div class="layout-page">
    <div class="row">
      <div class="col-md-3 col-ddm" data-ddm-field-row="0" data-ddm-field-page="0" data-ddm-field-column="0">
        <p>Radio Field</p>
      </div>
    </div>
    <div class="row">
      <div class="col-md-4 col-ddm" data-ddm-field-row="1" data-ddm-field-page="0" data-ddm-field-column="0">
        <p>Text Field</p>
      </div>
      <div class="col-md-6 col-ddm" data-ddm-field-row="1" data-ddm-field-page="0" data-ddm-field-column="1">
        <p>Select Field</p>
      </div>
    </div>
    <div class="row">
      <div class="col-md-12 col-ddm" data-ddm-field-row="2" data-ddm-field-page="0" data-ddm-field-column="0">
        <p>Text Field</p>
      </div>
    </div>
  </div>
</div>
`;

exports[`LayoutRenderer should render a layout with pages in mode of list 1`] = `
<div aria-labelledby="sidebarLightFormsFieldTab" class="fade tab-pane active show" id="sidebarLightDetails" role="tabpanel">
  <div class="row">
    <div class="col-md-3">
      <p>Radio Field</p>
    </div>
  </div>
  <div class="row">
    <div class="col-md-4">
      <p>Text Field</p>
    </div>
    <div class="col-md-6">
      <p>Select Field</p>
    </div>
  </div>
  <div class="row">
    <div class="col-md-12">
      <p>Text Field</p>
    </div>
  </div>
</div>
`;

exports[`LayoutRenderer should render default markup 1`] = `<p>Drag from sidebar and drop here</p>`;
